simplefs/simplefs.go

```diff
@@ -161,11 +214,130 @@ func (k *SimpleFS) favoriteList(ctx context.Context, path keybase1.Path, t tlf.T
 			k.log.Errorf("CanonicalToPreferredName: %q %v", fav.Name, err)
 			continue
 		}
-		res[string(pname)] = libkbfs.EntryInfo{Type: libkbfs.Dir}
+		res = append(res, keybase1.Dirent{})
+		res[len(res)-1].Name = string(pname)
+		res[len(res)-1].DirentType = deTy2Ty(libkbfs.Dir)
 	}
 	return res, nil
 }
 
+func setStat(de *keybase1.Dirent, fi os.FileInfo) {
+	de.Time = keybase1.ToTime(fi.ModTime())
+	de.Size = int(fi.Size()) // TODO: FIX protocol
+	t := libkbfs.File
+	if fi.IsDir() {
+		t = libkbfs.Dir
+	} else if fi.Mode()&0100 != 0 {
+		t = libkbfs.Exec
+	} else if fi.Mode()&os.ModeSymlink != 0 {
+		t = libkbfs.Sym
+	}
+	de.DirentType = deTy2Ty(t)
+}
+
+func (k *SimpleFS) setResult(opid keybase1.OpID, val interface{}) {
+	k.lock.Lock()
+	k.handles[opid] = &handle{async: val}
+	k.lock.Unlock()
+}
+
+func (k *SimpleFS) startOp(ctx context.Context, opid keybase1.OpID,
+	desc keybase1.OpDescription) (context.Context, error) {
+	ctx = k.makeContext(ctx)
+	ctx, cancel := context.WithCancel(ctx)
+	k.lock.Lock()
+	k.inProgress[opid] = &inprogress{desc, cancel, make(chan error, 1)}
+	k.lock.Unlock()
+	// ignore error, this is just for logging.
+	descBS, _ := json.Marshal(desc)
+	k.log.CDebugf(ctx, "start %X %s", opid, descBS)
+	return k.startOpWrapContext(ctx)
+}
+
+func (k *SimpleFS) doneOp(ctx context.Context, opid keybase1.OpID, err error) {
+	k.lock.Lock()
+	w, ok := k.inProgress[opid]
```

Comment:
> I might be missing something, but it seems we are missing a delete(k.inProgress, opid) in this function. I don't see a delete in startAsync, so it could leak *inprogress right?

It seems there's an implicit requirement that the caller of every async function calls `SimpleFSWait()` as well.  This is what the CLI calls do.  If we want to change that for the GUI work, we'll need to refactor this code a bunch I guess.  Should probably be a new ticket if needed.

Reply:
Oh I see; I missed that the `SimpleFSWait` removes it from the map, and that explain why it needs to remain in the map even after the operation is done. Thanks!

I don't have a reason for changing this behavior, and it seems OK to me.